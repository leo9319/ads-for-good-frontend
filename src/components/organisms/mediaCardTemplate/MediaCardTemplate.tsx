import React, { ReactElement } from 'react';
import VideoPlayer from '@components/atoms/VideoPlayer';
import Image from '@components/atoms/Image';
import Box from '@radix-styles/atoms/Box';
import Card from '@radix-styles/atoms/Card';
import Flex from '@radix-styles/atoms/Flex';
import Text from '@radix-styles/atoms/Text';
import { classNames } from '@utils/common/classNames';
import FallbackImage from '@assets/images/FallbackImage.svg';
import styles from './MediaCardTemplate.module.scss';
import { AspectRatioType } from '@radix-styles/atoms/AspectRatio/AspectRatio';

/**
 * @interface
 * Defines the common properties shared by all media types
 * This serves as a base interface for both video and image media
 */
interface MediaProps {
  /**
   * Defines the type of media to be displayed
   * @property
   */
  mediaType: 'image' | 'video';
  /**
   * The date associated with the media content that
   * used for displaying publication or event dates
   * @property
   */
  date?: string;
  /**
   * The title of the media content
   * @property
   */
  title: string;
  /**
   * A brief summary or description of the media content
   * @property
   */
  subHead?: string;
  /**
   * Fallback image Source for the image, this is will work if actual image is broken
   * @property
   */
  imageFallbackThumbnailSrc?: string;
  /**
   * Alternate text for the image
   * @property
   */
  imageAltText?: string;
  /**
   * An optional parameter to modify the existing card styles
   * @property
   */
  cardClassName?: string;
  /**
   * An optional paramenter to modify the existing content container styles
   * @property
   */
  contentContainerClass?: string;
  /**
   * An optional paramenter to change the image aspect ratio
   * @property
   */
  imgAspectRatio?: AspectRatioType;
  /**
   * Size variant of the component.
   * '1', '2', or '3' apply specific spacing or layout rules.
   * 'none' disables size-related styling.
   */
  size?: '1' | '2' | '3' | 'none';
  /**
   * If true, the component is rendered in article layout mode
   */
  isArticle?: boolean;
  /**
   * Click event handler for the component.
   */
  onClick?: (pageName?: string | null, id?: string) => void;
  /**
   * Unique ID of the card
   */
  id: string;
  /**
   * Unique ID to generate the slug
   */
  pageName?: string | null;
}

/**
 * @interface
 * Represents properties specific to video media
 */
export interface VideoMediaProps extends MediaProps {
  /**
   * Specifies that this media is a video
   * @property
   */
  mediaType: 'video';
  /**
   * The source URL of the video
   * This field is required for video media types
   * @property
   */
  videoSrc: string;
  /**
   * An optional thumbnail image for the video
   * Useful for displaying a preview before the video plays
   * @property
   */
  thumbnail?: string;
  /**
   * Prevents `imageSrc` from being used in video media
   * @property
   */
  imageSrc?: never;
}

/**
 * @interface
 * Represents properties specific to image media
 */
export interface ImageMediaProps extends MediaProps {
  /**
   * Specifies that this media is an image
   * @property
   */
  mediaType: 'image';
  /**
   * The source URL of the image
   * @property
   */
  imageSrc: string;
  /**
   * Prevents `videoSrc` from being used in image media
   * @property
   */
  videoSrc?: never;
  /**
   * Prevents `thumbnail` from being used in image media
   * @property
   */
  thumbnail?: never;
}

/**
 * @interface
 * A union type representing both image and video media properties
 * Ensures that only one type of media (either video or image) is used at a time
 */
export type MediaCardTemplateProps = ImageMediaProps | VideoMediaProps;

/**
 * The Template component for both Image and Video Media Cards
 *
 * @group Organisms
 * @category Component
 *
 * @param {MediaCardTemplateProps} props - The Media Card props
 * @returns {ReactElement} - Media Card component
 *
 */
export const MediaCardTemplate = ({
  date,
  title,
  subHead,
  mediaType,
  thumbnail,
  cardClassName,
  contentContainerClass,
  videoSrc,
  imageSrc,
  size,
  onClick,
  imageAltText = 'Media Card',
  imageFallbackThumbnailSrc = FallbackImage,
  imgAspectRatio = 'none',
  isArticle = false,
  id,
  pageName,
}: MediaCardTemplateProps): ReactElement => {
  const showImageMedia = mediaType === 'image';
  const articleClass = isArticle ? styles.article : '';
  const articleSize = isArticle && size ? styles[`size-${size}`] : '';

  const handleClick = (pageName?: string | null, id?: string) => {
    if (onClick) {
      onClick(pageName, id);
    }
  };

  return (
    <Card
      className={classNames(
        styles.card,
        articleClass,
        articleSize,
        cardClassName
      )}
      key={id}
      id={id}
      onClick={isArticle ? () => handleClick(pageName, id) : () => {}}
    >
      <Flex display="flex" className={styles.container}>
        <Box className={classNames(styles.mediaContainer, articleSize)}>
          {showImageMedia ? (
            <Image
              src={imageSrc}
              alt={imageAltText}
              fallbackSrc={imageFallbackThumbnailSrc}
              className={styles.imagePlaceholder}
              aspectRatio={imgAspectRatio}
            />
          ) : (
            <VideoPlayer
              size="none"
              src={videoSrc}
              thumbnail={thumbnail}
              containerClassName={styles.videoContainer}
              screenContainerClassName={styles.videoPlaceholder}
            />
          )}
        </Box>
        <Box
          className={classNames(
            styles.boxContainer,
            articleClass,
            articleSize,
            contentContainerClass
          )}
        >
          <Box className={styles.contentContainer}>
            {date && (
              <Text as="p" className={styles.date}>
                {date}
              </Text>
            )}
            <Box className={styles.contents}>
              {title && (
                <Text
                  as="p"
                  className={classNames(
                    styles.title,
                    articleClass,
                    articleSize
                  )}
                >
                  {title}
                </Text>
              )}
              {subHead && (
                <Text
                  as="p"
                  className={classNames(
                    styles.subHead,
                    articleClass,
                    articleSize
                  )}
                >
                  {subHead}
                </Text>
              )}
            </Box>
          </Box>
        </Box>
      </Flex>
    </Card>
  );
};

export default MediaCardTemplate;
