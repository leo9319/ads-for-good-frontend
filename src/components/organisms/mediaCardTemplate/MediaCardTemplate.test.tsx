import React from 'react';
import { render } from '@testing-library/react';
import MediaCardTemplate from './MediaCardTemplate';

describe('Media Card', () => {
  it('Video Media Card should render successfully', () => {
    const { baseElement } = render(
      <MediaCardTemplate
        date="Mar 4, 2025"
        mediaType="video"
        id="sample-123"
        videoSrc="https://www.youtube.com/embed/J2ogs029GVg?si=GRnGAS3WvToc-CXL"
        subHead="Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus."
        thumbnail="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
        title="Video: Burundi schools serve up success one plate at a time"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Image Media Card should render successfully', () => {
    const { baseElement } = render(
      <MediaCardTemplate
        mediaType="image"
        date="Mar 4, 2025"
        id="sample-456"
        imageSrc="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
        subHead="Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus."
        title="Video: Burundi schools serve up success one plate at a time"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Article Card should render successfully', () => {
    const { baseElement } = render(
      <MediaCardTemplate
        mediaType="image"
        date="Mar 4, 2025"
        id="sample-678"
        imageSrc="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
        subHead="Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus."
        title="Video: Burundi schools serve up success one plate at a time"
        isArticle
        size="2"
        imgAspectRatio="1:1"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
