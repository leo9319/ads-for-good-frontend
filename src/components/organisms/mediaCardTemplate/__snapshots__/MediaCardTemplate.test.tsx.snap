// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Card Article Card should render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface card article size-2"
      id="sample-678"
    >
      <div
        class="rt-Flex rt-r-display-flex container"
      >
        <div
          class="rt-Box mediaContainer size-2"
        >
          <div
            data-radix-aspect-ratio-wrapper=""
            style="position: relative; width: 100%; padding-bottom: 100%;"
          >
            <div
              style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
            >
              <img
                alt="Media Card"
                class="imgWrapper imagePlaceholder"
                src="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
                style="--imgWidth: 100%; --imgHeight: 100%;"
              />
            </div>
          </div>
        </div>
        <div
          class="rt-Box boxContainer article size-2"
        >
          <div
            class="rt-Box contentContainer"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default date"
              id=""
            >
              Mar 4, 2025
            </p>
            <div
              class="rt-Box contents"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title article size-2"
                id=""
              >
                Video: Burundi schools serve up success one plate at a time
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subHead article size-2"
                id=""
              >
                Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Media Card Image Media Card should render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface card"
      id="sample-456"
    >
      <div
        class="rt-Flex rt-r-display-flex container"
      >
        <div
          class="rt-Box mediaContainer"
        >
          <img
            alt="Media Card"
            class="imgWrapper imagePlaceholder"
            src="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
          />
        </div>
        <div
          class="rt-Box boxContainer"
        >
          <div
            class="rt-Box contentContainer"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default date"
              id=""
            >
              Mar 4, 2025
            </p>
            <div
              class="rt-Box contents"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                Video: Burundi schools serve up success one plate at a time
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subHead"
                id=""
              >
                Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Media Card Video Media Card should render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface card"
      id="sample-123"
    >
      <div
        class="rt-Flex rt-r-display-flex container"
      >
        <div
          class="rt-Box mediaContainer"
        >
          <div
            class="videoContainer videoContainer"
          >
            <div
              class="imageContainer"
            >
              <button
                class="buttonContainer"
              >
                <img
                  alt="Play Button"
                  class="imgWrapper playButton button none"
                  src="SvgrURL"
                />
                <img
                  alt="Video Player"
                  class="imgWrapper videoPlaceholder none videoPlaceholder"
                  src="https://images.pexels.com/photos/2113709/pexels-photo-2113709.jpeg"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="rt-Box boxContainer"
        >
          <div
            class="rt-Box contentContainer"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default date"
              id=""
            >
              Mar 4, 2025
            </p>
            <div
              class="rt-Box contents"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                Video: Burundi schools serve up success one plate at a time
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subHead"
                id=""
              >
                Proin congue lorem sit amet augue feugiat, eu placerat odio interdum. Mauris tincidunt arcu ut justo sollicitudin, eget volutpat magna ultrices. Nam tempus dapibus vehicula. Morbi blandit molestie metus vitae rhoncus. In in turpis sed ante euismod congue. Duis eu risus a nisl dapibus ultricies. Sed ultrices auctor lorem a finibus.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
